File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappData.java
package net.corda.plugins.cpk2;

import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Optional;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;

public class CordappData {
    private final Property<String> name;
    private final Property<Integer> versionId;
    private final Property<String> vendor;
    private final Property<String> licence;
    private final Property<String> cordappName;

    @Inject
    public CordappData(@NotNull ObjectFactory objects) {
        name = objects.property(String.class);
        versionId = objects.property(Integer.class);
        vendor = objects.property(String.class);
        licence = objects.property(String.class);
        cordappName = objects.property(String.class);
    }

    @Optional
    @Input
    public Property<String> getName() {
        return name;
    }

    @Optional
    @Input
    public Property<Integer> getVersionId() {
        return versionId;
    }

    @Optional
    @Input
    public Property<String> getVendor() {
        return vendor;
    }

    @Optional
    @Input
    public Property<String> getLicence() {
        return licence;
    }

    @Optional
    @Input
    public Property<String> getCordappName() {
        return cordappName;
    }

    boolean isEmpty() {
        return (!name.isPresent() && !versionId.isPresent() && !vendor.isPresent() && !licence.isPresent());
    }

    public void name(@Nullable String value) {
        name.set(value);
    }

    public void versionId(@Nullable Integer value) {
        versionId.set(value);
    }

    public void vendor(@Nullable String value) {
        vendor.set(value);
    }

    public void licence(@Nullable String value) {
        licence.set(value);
    }
}

File: cordapp-cpk2/src/main/kotlin/net/corda/plugins/cpk2/CordappData.kt
@file:JvmName("CordappDataProperties")
package net.corda.plugins.cpk2

import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.TaskInputs
import javax.inject.Inject

/**
 * Registers these [CordappData] properties as task inputs,
 * because Gradle cannot "see" their `@Input` annotations yet.
 */
fun TaskInputs.nested(nestName: String, data: CordappData) {
    property("${nestName}.name", data.name).optional(true)
    property("${nestName}.versionId", data.versionId).optional(true)
    property("${nestName}.vendor", data.vendor).optional(true)
    property("${nestName}.licence", data.licence).optional(true)
}

@Suppress("Unused")
open class CordappData @Inject constructor(objects: ObjectFactory) {
    @get:Optional
    @get:Input
    val name: Property<String> = objects.property(String::class.java)

    @get:Optional
    @get:Input
    val versionId: Property<Int> = objects.property(Int::class.java)

    @get:Optional
    @get:Input
    val vendor: Property<String> = objects.property(String::class.java)

    @get:Optional
    @get:Input
    val licence: Property<String> = objects.property(String::class.java)

    internal val isEmpty: Boolean
        get() = (!name.isPresent && !versionId.isPresent && !vendor.isPresent && !licence.isPresent)

    fun name(value: String?) {
        name.set(value)
    }

    fun versionId(value: Int?) {
        versionId.set(value)
    }

    fun vendor(value: String?) {
        vendor.set(value)
    }

    fun licence(value: String?) {
        licence.set(value)
    }
}
